File: tree-search-bench/deepest-node.rb
#!/usr/bin/env ruby

require 'json'

def deepestNode(root)
  def do_deep_rec(n, depth)
    if !n
      return [nil, 0]
    end
    if !n["left"] and !n["right"]
      return [n, depth]
    end

    left = do_deep_rec(n["left"], depth + 1)
    right = do_deep_rec(n["right"], depth + 1)

    if (left[1] > right[1])
      return left
    end
    return right
  end
  do_deep_rec(root, 1)[0]
end

def time_func()
  starting = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  yield
  ending = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  ending - starting
end

tree = JSON.parse(File.read(ARGV[0]))

result = nil
elapsed = time_func {
  result = deepestNode(tree)
}
puts "deepestNode: #{elapsed*1000}ms"

$stderr.puts(result.to_json)

File: tree-search-bench/typescript/src/deepest-node.ts
#!/usr/bin/env node

// I use `node` above instead of `ts-node` because this will be compiled to JS before running.

import { readFileSync } from 'fs';

/**
 * Given the root of a binary tree, return a deepest node. For example, in the following tree, return d.
 *
 *     a
 *   / \
 *  b  c
 * /
 * d
 *
 */

interface Node {
  left?: Node;
  right?: Node;
  // tslint:disable-next-line: no-any
  val: any;
}

export const deepestNode = (root: Node): Node => {
  const doDeepRec = (
    n: Node | undefined,
    depth: number
  ): [Node | undefined, number] => {
    if (!n) {
      return [undefined, 0];
    }
    if (!n.left && !n.right) {
      return [n, depth];
    }

    const [left, dleft] = doDeepRec(n.left, depth + 1);
    const [right, dright] = doDeepRec(n.right, depth + 1);

    if (dleft > dright) {
      return [left, dleft];
    }
    return [right, dright];
  };
  return doDeepRec(root, 1)[0] as Node;
};

const inputPath = process.argv[2];
const tree = JSON.parse(String(readFileSync(inputPath)));

console.time('deepestNode');
const result = deepestNode(tree);
console.timeEnd('deepestNode');

console.error(JSON.stringify(result));
